### ISSUE TITLE
Dependency conflict when running in Azure Cloud Service

### ISSUE THREAD

**Issue by binkiesjoris** [Azure/azure-event-hubs-dotnet#382]
I updated from the old Service bus package to the new Eventhubs package, but now Cloud services don't work (when deployed), I get the following error:

`Could not load file or assembly 'System.Diagnostics.DiagnosticSource, Version=4.0.3.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)`

I think that Cloud services (by default, even if you turn it off), include Application Insights code, which has a dependency on `System.Diagnostics.DiagnosticSource, Version=4.0.2.0` (I found package 4.4.0 of the diagnostics lib included when remoting into the VM). I tried everything I could think of (adding a binding redirect, downgrading specific packages, installing pre-release versions, etc) to resolve the conflict but nothing worked, either the dll cant be found, or there are problems with the assembly manifests.

The worst part is that it works fine locally, but not once deployed making it very hard to debug. Is this something that can be fixed on your end or with some settings? Or do I have no choice but to keep using the other service bus packages?

**Comment by serkantkaraca** [comment]
Can you check what library/assembly is attempting to load Version=4.0.3.1?

Our dependency is >= 4.5.0. Apparently one of the packages deployed isn't following general dependency guidance.

**Comment by binkiesjoris** [comment]
Well that specific version is being loaded by the event hubs package, your dependency is package version >= 4.5.0 but that package contains a dll with version 4.0.3.1, package version 4.4.0 contains a dll with version 4.0.2.0 for instance. The `The located assembly's manifest definition does not match the assembly reference.` error stems from the fact that the fusion assembly resolver finds 2 versions of the dll, with different versions that don't match the manifest.

Nothing else in my codebase contains anything that has a dependency on `System.Diagnostics.DiagnosticSource` (the fusion log didn't specify the source assembly causing the error), but I remoted into the VM again and in the `../sitesroots/App_Data/packages` directory the following packages are present (basically a bunch of telemetry stuff added by Azure AND `System.Diagnostics.DiagnosticSource Version=4.4.0`:

https://i.ibb.co/jHqJtYL/Capture3.png

So I'm guessing this is not actually a problem on your end, but most likely something to do with the telemetry/insights packages having a incorrectly configured dependency on 4.4.0?

**Comment by serkantkaraca** [comment]
Correct, this is not on our end. Whatever package is causing the issue, that has either of below issues for package reference:
* Reference an exact version
* Reference with an upper version limit

Here is a good read for building healthy package dependencies - https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/dependencies
